File: frontend/src/types/status.ts
import type { EntityConfig } from './domain';

export type ManifestState = 'draft' | 'submitted' | 'in_transit' | 'received' | 'rejected';
export const ALL_MANIFEST_STATE: readonly ManifestState[] = ['draft', 'submitted', 'in_transit', 'received', 'rejected'];
export type CheckState = 'pending' | 'pass' | 'fail' | 'escalated';
export const ALL_CHECK_STATE: readonly CheckState[] = ['pending', 'pass', 'fail', 'escalated'];

export const ENTITY_CONFIGS: readonly EntityConfig[] = [
  { key: 'wasteGenerator', path: 'generators', label: '产废单位', statuses: ['active', 'restricted', 'suspended', 'expired'] as const, transitionRole: 'reviewer' },
  { key: 'carrierProfile', path: 'carriers', label: '承运资质', statuses: ['pending', 'verified', 'restricted', 'expired'] as const, transitionRole: 'reviewer' },
  { key: 'transferManifest', path: 'manifests', label: '转运清单', statuses: ['draft', 'submitted', 'in_transit', 'received', 'rejected'] as const, transitionRole: 'operator' },
  { key: 'complianceCheck', path: 'checks', label: '合规核验', statuses: ['pending', 'pass', 'fail', 'escalated'] as const, transitionRole: 'reviewer' }
];

export const TRANSITIONS: Readonly<Record<string, Readonly<Record<string, readonly string[]>>>> = {
  wasteGenerator: { active: ['restricted', 'suspended', 'expired'], restricted: ['active', 'suspended', 'expired'], suspended: ['active', 'expired'], expired: [] },
  carrierProfile: { pending: ['verified', 'restricted', 'expired'], verified: ['restricted', 'expired'], restricted: ['pending', 'expired'], expired: [] },
  transferManifest: { draft: ['submitted'], submitted: ['in_transit', 'rejected'], in_transit: ['received', 'rejected'], received: [], rejected: [] },
  complianceCheck: { pending: ['pass', 'fail', 'escalated'], pass: [], fail: ['escalated'], escalated: [] }
};

File: backend/internal/constants/status.go
package constants

// Shared status values are mirrored in frontend/src/types/status.ts. Keeping
// the lists explicit makes state-machine drift visible during code review.

type ManifestState string

const (
	ManifestStateDraft     ManifestState = "draft"
	ManifestStateSubmitted ManifestState = "submitted"
	ManifestStateInTransit ManifestState = "in_transit"
	ManifestStateReceived  ManifestState = "received"
	ManifestStateRejected  ManifestState = "rejected"
)

var AllManifestState = []string{"draft", "submitted", "in_transit", "received", "rejected"}

type CheckState string

const (
	CheckStatePending   CheckState = "pending"
	CheckStatePass      CheckState = "pass"
	CheckStateFail      CheckState = "fail"
	CheckStateEscalated CheckState = "escalated"
)

var AllCheckState = []string{"pending", "pass", "fail", "escalated"}

var WasteGeneratorTransitions = map[string]map[string]bool{
	"active":     {"restricted": true, "suspended": true, "expired": true},
	"restricted": {"active": true, "suspended": true, "expired": true},
	"suspended":  {"active": true, "expired": true},
	"expired":    {},
}

var CarrierProfileTransitions = map[string]map[string]bool{
	"pending":    {"verified": true, "restricted": true},
	"verified":   {"restricted": true, "expired": true},
	"restricted": {"pending": true, "expired": true},
	"expired":    {},
}

var TransferManifestTransitions = map[string]map[string]bool{
	"draft":      {"submitted": true},
	"submitted":  {"in_transit": true, "rejected": true},
	"in_transit": {"received": true, "rejected": true},
	"received":   {},
	"rejected":   {},
}

var ComplianceCheckTransitions = map[string]map[string]bool{
	"pending":   {"pass": true, "fail": true},
	"pass":      {},
	"fail":      {"escalated": true},
	"escalated": {},
}

func CanTransition(graph map[string]map[string]bool, from, to string) bool {
	targets, exists := graph[from]
	return exists && targets[to]
}
